### PR TITLE
CSE (v2)

### PR DESCRIPTION
**Pull request by vincepandolfo**
The problem with this one seems to be a loss of precision in the results. The tests that fail do so for that reason. I'm opening a PR anyway because I can't really pinpoint the problem by myself.

It needs to be integrated with `Iteration` objects as well but it's probably better to wait for #92 for that.

### CODE REVIEW

File: devito/propagator.py [hunk 1]
```diff
@@ -267,6 +267,22 @@ def sympy_to_cgen(self, stencils):
...
+        dtypes = []
+
+        for e in preorder_traversal(expr):
+            if hasattr(e, 'dtype'):
+                dtypes.append(e.dtype)
+
+        if dtypes:
+            return np.find_common_type(dtypes, [])
+        else:
+            return np.float64
```
Comment (mlange05) [inline devito/propagator.py]: Does the Propagator not have a `dtype` argument? Should we not be defaulting to this one?

File: devito/propagator.py [hunk 2]
```diff
@@ -282,12 +298,16 @@ def convert_equality_to_cgen(self, equality):
             s_lhs = ccode(self.time_substitutions(equality.lhs).xreplace(self._var_map))
             s_rhs = self.time_substitutions(equality.rhs).xreplace(self._var_map)
 
+            if s_lhs.find("temp") != -1:
+                s_lhs = "double" + " " + s_lhs
```
Comment (mlange05) [inline devito/propagator.py]: Is this hard-coding the type of temps? Are we not deriving their type above?

File: devito/propagator.py (hunk 1 above)
Comment (vincepandolfo) [inline devito/propagator.py]: Yes, we should. I overlooked that argument.

File: devito/propagator.py (hunk 2 above)
Comment (vincepandolfo) [inline devito/propagator.py]: I couldn't find a way to convert the dtype to the proper string easily so I left it this way while working on the actual CSE.

File: devito/operator.py [hunk 3]
```diff
@@ -62,6 +63,64 @@ def expr_indexify(expr):
     return expr.xreplace(replacements)
 
 
+def expr_cse(expr):
+    """Performs common subexpression elimination on expressio
```
Comment (FabioLuporini) [inline devito/operator.py]: typo

**Comment by FabioLuporini** [comment]
Great stuff!

I think we should rename ``factorized`` as ``temporaries`` or something else, but definitely not ``factorized``

I think it's time to move all functions/methods that visit and manipulate sympy expressions into a proper class (proper classes) in a separate module (``expressions.py`` ?) Not sure if this should come with this PR, but it should happen soon!

**Comment by mlange05** [comment]
@FabioLuporini Just FYI, PR #92 moves all expression traversal into a separate `Expression` class to aggregate them. In it's current state this is only done on the low-level "indexed" expressions, but it could be extended to cover all expression types.

**Comment by FabioLuporini** [comment]
@mlange05 I see, good

**Comment by mlange05** [comment]
One more thing; is CSE always performed or should we have a flag to swtich it on.off (default on, of course)?

**Comment by FabioLuporini** [comment]
I think we should have a flag at least for performance comparison (code gen time, and also performance of the generated code with different backend compilers)

**Comment by vincepandolfo** [comment]
Yes, that sounds like a sensible idea. Should we integrate this with `Iteration` already or is it better to wait for #92 to land?

**Comment by mlange05** [comment]
Let's leave the integration with `Iteration` for another PR and land this one first. I think we want the flag though.

**Comment by FabioLuporini** [comment]
Assuming all tests pass -- and after rebasing -- is it correct to say that this PR can land ?

**Comment by vincepandolfo** [comment]
Right now all the tests are running with `cse` off since the default value for the flag is `False`. Should I add more tests that use `cse` or change the default value?

**Comment by mloubout** [comment]
I think a test checking with/without cse produce the same result would be nice.

**Comment by FabioLuporini** [comment]
@vincepandolfo remind me why the flag is off by default? shouldn't it be on?

I agree with @mloubout : we should have a few tests, that we know stress cse quite hard, to check the numerical correctness

**Comment by vincepandolfo** [comment]
The flag is on by default now

File: tests/test_cse.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    receiver_coords = np.zeros((101, 3))
+    receiver_coords[:, 0] = np.linspace(50, 950, num=101)
+    receiver_coords[:, 1] = 500
+    receiver_coords[:, 2] = location[2]
+    data.set_receiver_pos(receiver_coords)
+    data.set_shape(nt, 101)
+    Acoustic = Acoustic_cg(model, data)
+    (rec, u) = Acoustic.Forward(save=True, cse=cse)
+
```
Comment (mloubout) [inline tests/test_cse.py]: `save=True` is not necessary unless you want to compare the actual wavefield with
`assert np.isclose(np.linalg.norm(u- u_cse), 0.0)`
If not set to `False` as Travis really don't like large arrays

Comment (vincepandolfo) [inline tests/test_cse.py]: Good to know. Will change now

**Pull request merged**